### PR TITLE
Http transport rate limits

### PR DESCRIPTION
**Pull request by adovh**
## Pull Request description

[Jira](https://thingsboard-portal.atlassian.net/browse/PROD-1333)

## General checklist

- [x] You have reviewed the guidelines [document](https://docs.google.com/document/d/1wqcOafLx5hth8SAg4dqV_LV3un3m5WYR8RdTJ4MbbUM/edit?usp=sharing).
- [x] [Labels](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/managing-labels#about-labels) that classify your pull request have been added.
- [x] The [milestone](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/about-milestones) is specified and corresponds to fix version.  
- [x] Description references specific [issue](https://github.com/thingsboard/thingsboard/issues).
- [x] Description contains human-readable scope of changes.
- [x] Description contains brief notes about what needs to be added to the documentation.
- [x] No merge conflicts, commented blocks of code, code formatting issues.
- [x] Changes are backward compatible or upgrade script is provided.
- [x] Similar PR is opened for PE version to simplify merge. Crosslinks between PRs added. Required for internal contributors only.
  
## Front-End feature checklist

- [ ] Screenshots with affected component(s) are added. The best option is to provide 2 screens: before and after changes;
- [ ] If you change the widget or other API, ensure it is backward-compatible or upgrade script is present.
- [ ] Ensure new API is documented [here](https://github.com/thingsboard/thingsboard-ui-help)

## Back-End feature checklist

- [x] Added corresponding unit and/or integration test(s). Provide written explanation in the PR description if you have failed to add tests.
- [x] If new dependency was added: the dependency tree is checked for conflicts.
- [x] If new service was added: the service is marked with corresponding @TbCoreComponent, @TbRuleEngineComponent, @TbTransportComponent, etc.
- [x] If new REST API was added: the RestClient.java was updated, issue for [Python REST client](https://github.com/thingsboard/thingsboard-python-rest-client) is created.

### CODE REVIEW

File: common/transport/http/src/main/java/org/thingsboard/server/transport/http/limits/IpRateLimitInterceptor.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    }
+    @Override
+    public boolean preHandle(HttpServletRequest request,
+                             HttpServletResponse response,
+                             Object handler) throws Exception {
+
+        if (ctx.checkAddress(getInetSocketAddress(request))){
+            return true;
+        } else {
+            response.setStatus(401);
```
Comment (YevhenBondarenko) [inline common/transport/http/src/main/java/org/thingsboard/server/transport/http/limits/IpRateLimitInterceptor.java:46]: Use constant `HttpStatus.UNAUTHORIZED`

File: application/src/test/java/org/thingsboard/server/system/BaseHttpDeviceApiTest.java [hunk 2]
```diff
@@ -60,6 +67,22 @@ public void before() throws Exception {
...
+    public void testRateLimits_wrongCredentialsMaxOut() throws Exception {
+        doGetAsync("/api/v1/" + deviceCredentials.getCredentialsId() + "/attributes?clientKeys=keyA,keyB,keyC")
+                .andExpect(status().isOk());
+
+        for (int i = 0; i < maxWrongCredentialsPerIp; i++){
+            doGetAsync("/api/v1/" + "WRONG_TOKEN" + "/attributes?clientKeys=keyA,keyB,keyC")
+                    .andExpect(status().isUnauthorized());
+        }
+        mockMvc.perform(get("/api/v1/" + deviceCredentials.getCredentialsId() + "/attributes?clientKeys=keyA,keyB,keyC"))
+                .andExpect(status().is4xxClientError());
```
Comment (YevhenBondarenko) [inline application/src/test/java/org/thingsboard/server/system/BaseHttpDeviceApiTest.java:80]: we should check 401

File: application/src/test/java/org/thingsboard/server/system/BaseHttpDeviceApiTest.java [hunk 3]
```diff
@@ -60,6 +67,22 @@ public void before() throws Exception {
...
+        Thread.sleep(ipBlockTimeout);
```
Comment (YevhenBondarenko) [inline application/src/test/java/org/thingsboard/server/system/BaseHttpDeviceApiTest.java:81]: TransportService.process should not be invoked after IP blocking
plz check it

File: application/src/test/java/org/thingsboard/server/system/BaseHttpDeviceApiTest.java [hunk 4]
```diff
@@ -60,6 +67,22 @@ public void before() throws Exception {
                 doGet("/api/device/" + device.getId().getId().toString() + "/credentials", DeviceCredentials.class);
     }
 
+    @Test
+    public void testRateLimits_wrongCredentialsMaxOut() throws Exception {
+        doGetAsync("/api/v1/" + deviceCredentials.getCredentialsId() + "/attributes?clientKeys=keyA,keyB,keyC")
+                .andExpect(status().isOk());
+
+        for (int i = 0; i < maxWrongCredentialsPerIp; i++){
+            doGetAsync("/api/v1/" + "WRONG_TOKEN" + "/attributes?clientKeys=keyA,keyB,keyC")
```
Comment (YevhenBondarenko) [inline application/src/test/java/org/thingsboard/server/system/BaseHttpDeviceApiTest.java:76]: TransportService.process should be invoked `maxWrongCredentialsPerIp` times
plz check it

File: common/transport/http/src/main/java/org/thingsboard/server/transport/http/limits/IpRateLimitInterceptor.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+            ctx.onAuthSuccess(addr);
+        } else {
+            ctx.onAuthFailure(addr);
+        }
+    }
+
+    @NotNull
+    private InetSocketAddress getInetSocketAddress(HttpServletRequest request) {
+        String addr = ctx.isProxyEnabled()? request.getHeaders("X-Forwarded-For").nextElement() : request.getRemoteAddr();
+        return new InetSocketAddress(addr, 80);
```
Comment (YevhenBondarenko) [inline common/transport/http/src/main/java/org/thingsboard/server/transport/http/limits/IpRateLimitInterceptor.java:65]: we use one port for all clients from one IP, it's not ok
use `request.getRemotePort()`

**Review by YevhenBondarenko — APPROVED** [review]

File: common/transport/http/src/main/java/org/thingsboard/server/transport/http/limits/IpRateLimitInterceptor.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+        if (status != HttpStatus.UNAUTHORIZED.value()){
+            ctx.onAuthSuccess(addr);
+        } else {
+            ctx.onAuthFailure(addr);
+        }
+    }
+
+    @NotNull
+    private InetSocketAddress getInetSocketAddress(HttpServletRequest request) {
+        String addr = ctx.isProxyEnabled()? request.getHeaders("X-Forwarded-For").nextElement() : request.getRemoteAddr();
```
Comment (ashvayka) [inline common/transport/http/src/main/java/org/thingsboard/server/transport/http/limits/IpRateLimitInterceptor.java:64]: We should just check for ```X-Forwarded-For```, and we DO NOT need ```HTTP_PROXY_PROTOCOL_ENABLED``` at all. All production deployments are under a load balancer. We should just check the presence of this header.

Comment (ashvayka) [reply common/transport/http/src/main/java/org/thingsboard/server/transport/http/limits/IpRateLimitInterceptor.java:64]: Extract the ```X-Forwarded-For``` into String constant and use it in the  ```RestAuthenticationDetails``` as well.

**Pull request closed without merging**